contracts/Tokenator.sol

```diff
@@ -0,0 +1,176 @@
+    /// @param totalSupply_ token total supply
+    /// @param underlyingToken_ the token that tokenator can be redeemed for
+    /// @param commencement_ timestamp at which tokenator vesting begins
+    /// @param durationDays_ duration in days for tokenator to fully vest
+    constructor(
+        string memory name_,
+        string memory symbol_,
+        uint256 totalSupply_,
+        IERC20 underlyingToken_,
+        uint256 commencement_,
+        uint256 durationDays_
+    ) ERC20(name_, symbol_) Ownable(msg.sender) {
+        if (commencement_ < block.timestamp) revert CommencementPassed();
+        if (address(underlyingToken_) == address(0)) revert ZeroAddress();
+
+        underlyingToken = underlyingToken_;
+        commencement = commencement_;
+        durationSeconds = durationDays_ * 1 days;
+
+        _mint(msg.sender, totalSupply_);
+
+        emit Deployed(
+            msg.sender,
+            underlyingToken_,
+            commencement_,
+            durationDays_
+        );
+    }
+
+    /// @notice allows user to claim underlying token from this contract by burning tokenator token
+    /// @param tokenatorAmount amount of tokenator token to burn
+    /// @param to address to receive the underlying token
+    /// @return underlyingTokenAmount amount of underlyingToken `to` address receives
+    function claimUnderlyingToken(
+        uint256 tokenatorAmount,
+        address to
+    ) external returns (uint256 underlyingTokenAmount) {
+        if (commencement >= block.timestamp) revert CommencementNotPassed();
+        if (to == address(0)) revert ZeroAddress();
+
+        underlyingTokenAmount = underlyingTokenForTokenator(tokenatorAmount);
+        _burn(msg.sender, tokenatorAmount);
+        underlyingToken.safeTransfer(to, underlyingTokenAmount);
+
+        emit ClaimUnderlyingToken(
+            msg.sender,
+            to,
+            tokenatorAmount,
+            underlyingTokenAmount
+        );
+    }
+
+    /// @notice owner may withdraw underlying token in emergency situations
+    /// @param amount underlying token amount to withdraw
+    /// @param to address to receive the underlying token
+    function emergencyWithdraw(uint256 amount, address to) external onlyOwner {
+        if (to == address(0)) revert ZeroAddress();
+
+        underlyingToken.safeTransfer(to, amount);
+
+        emit EmergencyWithdrawal(msg.sender, amount, to);
+    }
+
+    /// @notice owner may withdraw tokens accidentally sent to the contract
+    /// @notice this function can't be used to withdraw the underlying token
+    /// @param token the token to withdraw
+    /// @param amount the amount of the token to withdraw
+    /// @param to the address to receive the rescued token
+    function rescueTokens(
+        IERC20 token,
+        uint256 amount,
+        address to
+    ) external onlyOwner {
+        if (to == address(0)) revert ZeroAddress();
+        if (token == underlyingToken) revert CannotRescueUnderlyingToken();
```

Comment:
if you're going to have `emergencyWithdraw` that can rescue the `underlyingToken` then you might as well remove `emergencyWithdraw` and allow `rescueTokens` to rescue `underlyingToken`.